### PR TITLE
EVA-2722: downloading analysis (vcf file) from through ENA rest end point

### PR DESCRIPTION
**Pull request by nitin-ebi**

### CODE REVIEW

File: covid19dp_submission/steps/download_analysis.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+def main():
+    parser = argparse.ArgumentParser(description='Download analysis for processing from the Covid-19 DP project',
+                                     formatter_class=argparse.RawTextHelpFormatter, add_help=False)
+    parser.add_argument("--project", default='PRJEB45554', required=False,
+                        help="project from which analysis needs to be downloaded")
+    parser.add_argument("--num-analysis", type=int, default=10000, required=False,
+                        help="Number of analysis to download (max = 10000)")
+    parser.add_argument("--processed-analysis-file",
+                        default="/nfs/production3/eva/data/PRJEB45554/covid19_analysis.txt",
```
Comment (sundarvenkata-EBI) [inline covid19dp_submission/steps/download_analysis.py]: Suggest not using cluster paths in public repositories..

File: covid19dp_submission/steps/download_analysis.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                                                                :(num_analysis - len(analysis_for_processing))]
+            break
+        else:
+            analysis_for_processing = analysis_for_processing + unprocessed_analysis
+            offset = offset + limit
+
+    return analysis_for_processing
+
+
+def get_analysis_from_ena(project, offset, limit):
```
Comment (sundarvenkata-EBI) [inline covid19dp_submission/steps/download_analysis.py]: Suggest annotating with retry and jitter as suggested in the stand-up meeting today...

File: covid19dp_submission/steps/download_analysis.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+def get_processed_analysis(processed_analysis_file):
+    processed_analysis = set()
+    with open(processed_analysis_file, 'r') as file:
+        for line in file:
+            processed_analysis.add(line.split(",")[0])
+    return processed_analysis
+
+
+def download_files(analysis_array, download_target_dir, processed_analysis_file):
```
Comment (sundarvenkata-EBI) [inline covid19dp_submission/steps/download_analysis.py]: Suggest annotating with retry and jitter as suggested in the stand-up meeting today...

File: covid19dp_submission/steps/download_analysis.py (hunk 1 above)
Comment (sundarvenkata-EBI) [inline covid19dp_submission/steps/download_analysis.py]:
```suggestion
                        default="/nfs/production3/eva/data/PRJEB45554/covid19_processed_analyses.txt",
```

**Review by sundarvenkata-EBI — APPROVED** [review]
Looks good! Suggest adding tests or creating a separate ticket for that..

File: covid19dp_submission/steps/download_analysis.py (hunk 1 above)
Comment (nitin-ebi) [reply covid19dp_submission/steps/download_analysis.py]: removed default path and made the parameter mandatory

File: covid19dp_submission/steps/download_analysis.py (hunk 2 above)
Comment (nitin-ebi) [reply covid19dp_submission/steps/download_analysis.py]: added

File: covid19dp_submission/steps/download_analysis.py (hunk 3 above)
Comment (nitin-ebi) [reply covid19dp_submission/steps/download_analysis.py]: added.

**Review by sundarvenkata-EBI — APPROVED** [review]

File: covid19dp_submission/steps/download_analysis.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import urllib
+
+import requests
+from ebi_eva_common_pyutils.logger import logging_config
+from retry import retry
+
+logger = logging_config.get_logger(__name__)
+
+
+def download_analysis(project, num_analysis, processed_analysis_file, download_target_dir):
```
Comment (tcezard) [inline covid19dp_submission/steps/download_analysis.py]:
```suggestion
def download_analyses(project, num_analysis, processed_analysis_file, download_target_dir):
```

File: covid19dp_submission/steps/download_analysis.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                        help="full path to the file containing all the processed analysis")
+    parser.add_argument("--download-target-dir", required=True, help="Full path to the target download directory")
+
+    args = parser.parse_args()
+    logging_config.add_stdout_handler()
+
+    if args.num_analysis < 1 or args.num_analysis > 10000:
+        raise Exception("number of analysis to download should be between 1 and 10000")
+
+    download_analysis(args.project, args.num_analysis, args.processed_analysis_file, args.download_target_dir)
```
Comment (tcezard) [inline covid19dp_submission/steps/download_analysis.py]:
```suggestion
    download_analyses(args.project, args.num_analysis, args.processed_analysis_file, args.download_target_dir)
```

**Review by tcezard — APPROVED** [review]

File: covid19dp_submission/steps/download_analysis.py (hunk 4 above)
Comment (nitin-ebi) [reply covid19dp_submission/steps/download_analysis.py]: done

File: covid19dp_submission/steps/download_analysis.py (hunk 5 above)
Comment (nitin-ebi) [reply covid19dp_submission/steps/download_analysis.py]: done

**Review by sundarvenkata-EBI — APPROVED** [review]

File: tests/resources/processed_analyses_file.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+RZ3372540,ftp.sra.ebi.ac.uk/vol1/ERZ337/ERZ3372540/SRR15239121.vcf
```
Comment (tcezard) [inline tests/resources/processed_analyses_file.txt]: Is this missing the `E` at the start ?

File: tests/test_accession_and_clustering.py [hunk 7]
```diff
@@ -61,9 +60,17 @@ def tearDown(self) -> None:
         shutil.rmtree(self.processing_folder, ignore_errors=True)
         self.mongo_db.drop_database(self.accessioning_database_name)
 
+
+    def download_test_files(self):
+        download_file_name = os.path.basename(self.download_url)
+        snapshot_download_command = (f'bash -c "cd {self.download_target_dir} && curl -O {self.download_url} && '
```
Comment (tcezard) [inline tests/test_accession_and_clustering.py]: Not sure why we're using curl to retrieve test data. I know that's what was used before but if you're going to rewrite it I think it's unnecessary.

File: tests/resources/processed_analyses_file.txt (hunk 6 above)
Comment (nitin-ebi) [reply tests/resources/processed_analyses_file.txt]: Thanks, corrected

File: tests/test_accession_and_clustering.py (hunk 7 above)
Comment (nitin-ebi) [reply tests/test_accession_and_clustering.py]: refactored

**Pull request merged**